core/common/src/main/java/alluxio/concurrent/ManagedBlockingUfsForwarder.java

```diff
@@ -0,0 +1,654 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.concurrent;
+
+import alluxio.AlluxioURI;
+import alluxio.SyncInfo;
+import alluxio.collections.Pair;
+import alluxio.concurrent.jsr.ForkJoinPool;
+import alluxio.security.authorization.AccessControlList;
+import alluxio.security.authorization.AclEntry;
+import alluxio.security.authorization.DefaultAccessControlList;
+import alluxio.underfs.UfsDirectoryStatus;
+import alluxio.underfs.UfsFileStatus;
+import alluxio.underfs.UfsMode;
+import alluxio.underfs.UfsStatus;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.options.CreateOptions;
+import alluxio.underfs.options.DeleteOptions;
+import alluxio.underfs.options.FileLocationOptions;
+import alluxio.underfs.options.ListOptions;
+import alluxio.underfs.options.MkdirsOptions;
+import alluxio.underfs.options.OpenOptions;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Forwarder for {@link UnderFileSystem} objects that works through with ForkJoinPool's
+ * managed blocking.
+ *
+ * If UFS calls are being done on a {@link alluxio.concurrent.jsr.ForkJoinWorkerThread}, then
+ * this forwarder will make sure UFS operations are treated as blocking operations
+ * for compensating the ForkJoinPool.
+ *
+ */
+public class ManagedBlockingUfsForwarder implements UnderFileSystem {
+
+  /** Underlying {@link UnderFileSystem}. **/
+  private UnderFileSystem mUfs;
+
+  /**
+   * Creates {@link ManagedBlockingUfsForwarder} instance.
+   *
+   * @param ufs the underlying ufs
+   */
+  public ManagedBlockingUfsForwarder(UnderFileSystem ufs) {
+    mUfs = ufs;
+  }
+
+  @Override
+  public void cleanup() throws IOException {
+    mUfs.cleanup();
+  }
+
+  @Override
+  public void connectFromMaster(String hostname) throws IOException {
+    mUfs.connectFromMaster(hostname);
```

Comment:
Should this be wrapped in the managed blocker?

Reply:
What does this do? If it's slow and entails an UFS interaction, then yes.

Reply:
Take a look at the implementations of `connectFromMaster`. I think some may require UFS interaction but most do not.

Reply:
I see the slower UFSs don't even implement this method. So it should be fine to exclude them from managed blocking. Because managing blocking has its own overhead, which could make fast operations slower when used liberally.